File: lib/ssh-config.js
"use strict";

var _ = require('underscore')._;
var fs = require('fs');
var temp = require('temp');

function forwardConfig(fwds) {
    var lines = [];
    _.each(fwds, function (fs, descr) {
        lines.push('  # ' + descr);
        fs.forEach(function (f) {
            lines.push('  LocalForward ' + f.from + ' ' + f.to);
        });
    });
    return lines;
}

module.exports = function (config) {
    var lines = [];

    _.each(config.hosts, function (c, name) {
        lines.push('Host ' + name);
        lines.push('  Compression yes');

        if (c.user) {
            lines.push('  User ' + c.user);
        }
        if (c.addr) {
            lines.push('  Hostname ' + c.addr);
        }
        if (c.port) {
            lines.push('  Port ' + c.port);
        }
        if (c.via) {
            lines.push('  ProxyCommand ssh -F ' + config.sshConfig + ' ' + c.via + ' nc -w 1800 %h %p');
        }

        if (c.key) {
            var keyFile = temp.path({suffix: '.pem'});
            fs.writeFileSync(keyFile, c.key);
            fs.chmodSync(keyFile, 384 /* 0600 octal */);
            lines.push('  IdentityFile ' + keyFile);
            lines.push('  PubkeyAuthentication yes');
        } else {
            lines.push('  PubkeyAuthentication no');
        }

        if (c.password) {
            lines.push('  PasswordAuthentication yes');
            lines.push('  KbdInteractiveAuthentication yes');
        } else {
            lines.push('  PasswordAuthentication no');
            lines.push('  KbdInteractiveAuthentication no');
        }

        if (!c.keepalive) {
            c.keepalive = 180;
        }
        var countMax = Math.max(3, Math.floor(c.keepalive / 5));
        lines.push('  ServerAliveInterval 5');
        lines.push('  ServerAliveCountMax ' + countMax);

        if (name === config.main) {
            var forwards = forwardConfig(config.forwards);
            lines.push.apply(lines, forwards);
        }
    });

    return lines.join('\n') + '\n';
};

File: ssh-config.go
package main

import (
	"fmt"
	"strings"

	"github.com/calmh/mole/configuration"
	"github.com/calmh/mole/tmpfileset"
)

func sshConfig(cfg *configuration.Config, fs *tmpfileset.FileSet) {
	var lines []string
	nl := func(line string) {
		lines = append(lines, line)
	}

	for name, host := range cfg.Hosts {
		nl("Host " + name)
		nl("  HostName " + host.Addr)
		nl(fmt.Sprintf("  Port %d", host.Port))
		nl("  User " + host.User)

		if host.Via != "" {
			nl("  ProxyCommand ssh -F {ssh-config} " + host.Via + " nc -W 1800 %h %p")
		}

		nl("  KeychainIntegration no")
		if host.Pass != "" {
			nl("  # password")
			nl("  PasswordAuthentication yes")
			nl("  KbdInteractiveAuthentication yes")
			nl("  PubkeyAuthentication no")
			nl("  RSAAuthentication no")
		} else if host.Key != "" {
			fs.Add("identity-"+host.Unique, []byte(host.Key))
			nl("  # ssh key")
			nl("  IdentityFile {identity-" + host.Unique + "}")
			nl("  PasswordAuthentication no")
			nl("  KbdInteractiveAuthentication no")
			nl("  PubkeyAuthentication yes")
			nl("  RSAAuthentication yes")
		}

		nl("  Compression yes")
		nl("  # 60 second keepalive timeout")
		nl("  ServerAliveInterval 20")
		nl("  ServerAliveCountMax 3")

		if name == cfg.General.Main {
			for name, fwd := range cfg.Forwards {
				nl(fmt.Sprintf("  # forward %q", name))
				for _, l := range fwd.Lines {
					for i := 0; i <= l.Repeat; i++ {
						nl(fmt.Sprintf("  LocalForward %s:%d %s:%d", l.SrcIP, l.SrcPort+i, l.DstIP, l.DstPort+i))
					}
				}
			}
		}
	}

	content := []byte(strings.Join(lines, "\n") + "\n")
	fs.Add("ssh-config", content)
}
